### PR TITLE
Unit redirects should redirect all participants not just students

### PR DESCRIPTION
**Pull request by dmcavoy**
We currently have logic to redirect students to the course we think they should be. As we move toward a world where the people learning in a course can be more than students we need to expand that logic to redirect anyone who is a participant for the course. 

This PR makes it so we redirect anyone who is a participant in the course not just students. This means teachers in a course where participant_audience is teacher will get redirected in the same way we have been redirect students(considering most recent course, course where they have progress, course where they are assigned).

## Links

[Eng Plan](https://docs.google.com/document/d/1V61xONU0-mleMEEdHWNxhvZtYIPRCB31mT_hbUXusCQ/edit#)
[Product Spec](https://docs.google.com/document/d/1qZpc90daxQiiqpcQUiUiGV2tQoaYEQUzOA72UZzBlVE/edit#heading=h.f5x8no3lbpxy)

## Testing story

- Added new unit tests
- There are existing unit tests as well as UI tests for redirection
- We will be testing this logic as part of a bug bash before launching

### CODE REVIEW

File: dashboard/app/models/script.rb [hunk 1]
```diff
@@ -596,13 +596,25 @@ def self.get_unit_family_redirect_for_user(family_name, user: nil, locale: 'en-U
...
+    if user && family_units.last.can_be_participant?(user)
       assigned_unit_ids = user.section_scripts.pluck(:id)
       progress_unit_ids = user.user_levels.map(&:script_id)
       unit_ids = assigned_unit_ids.concat(progress_unit_ids).compact.uniq
       unit_name = family_units.select {|s| unit_ids.include?(s.id)}&.first&.name
-      return Script.new(redirect_to: unit_name, published_state: SharedCourseConstants::PUBLISHED_STATE.beta) if unit_name
+      if unit_name
+        # This creates a temporary script which is used to redirect the user. The audiences are set
+        # to instructor as teacher and student as participant because that will allow the redirect to
+        # happen for any course
```
Comment (davidsbailey) [inline dashboard/app/models/script.rb]: > The audiences are set to instructor as teacher and student as participant"

I'm having trouble understanding this part of the comment. I think the code says clearly enough what it is doing so I would leave this part out and just say why it is doing it, e.g. "The audiences are set to allow..."

I think I am missing something though. Why does this strategy allow redirects for any course? would it make more sense to say that this allows users of any type to be redirected?

File: dashboard/test/models/script_test.rb [hunk 2]
```diff
@@ -270,6 +298,17 @@ def populate_cache_and_disconnect_db
...
+    facilitator = create :facilitator
+    pl_csp1_2017 = create(:script, name: 'pl-csp1-2017', family_name: 'pl-csp', version_year: '2017', published_state: SharedCourseConstants::PUBLISHED_STATE.stable, instructor_audience: SharedCourseConstants::INSTRUCTOR_AUDIENCE.facilitator, participant_audience: SharedCourseConstants::PARTICIPANT_AUDIENCE.teacher)
+    pl_csp1_2018 = create(:script, name: 'pl-csp1-2018', family_name: 'pl-csp', version_year: '2018', published_state: SharedCourseConstants::PUBLISHED_STATE.stable, instructor_audience: SharedCourseConstants::INSTRUCTOR_AUDIENCE.facilitator, participant_audience: SharedCourseConstants::PARTICIPANT_AUDIENCE.teacher)
+    create(:script, name: 'pl-csp1-2019', family_name: 'pl-csp', version_year: '2019', instructor_audience: SharedCourseConstants::INSTRUCTOR_AUDIENCE.facilitator, participant_audience: SharedCourseConstants::PARTICIPANT_AUDIENCE.teacher)
+    create :section, user: facilitator, script: pl_csp1_2017
+
+    redirect_unit = Script.get_unit_family_redirect_for_user('pl-csp', user: facilitator)
+    assert_equal pl_csp1_2018.name, redirect_unit.redirect_to
+  end
+
```
Comment (davidsbailey) [inline dashboard/test/models/script_test.rb:311]: nice tests 👍

**Review by davidsbailey — APPROVED** [review]
just one issue with comments, otherwise looks good!

File: dashboard/app/models/script.rb (hunk 1 above)
Comment (dmcavoy) [reply dashboard/app/models/script.rb]: I have updated the comment based on your feedback. Thanks!

**Review by jamescodeorg — APPROVED** [review]
I'm not very familiar with this code, but it looks reasonable to me.

Side question, unrelated to your change: How often does this get called?  I'm wondering if this line: `progress_unit_ids = user.user_levels.map(&:script_id)` will query for all of user's user_levels on every call to `get_unit_family_redirect_for_user`.

**Comment by dmcavoy** [comment]
@jamescodeorg it looks like this method can get called frequently. I traced it all the way back to this line https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/scripts_controller.rb#L8 . So maybe something to look into?

**Comment by jamescodeorg** [comment]
> @jamescodeorg it looks like this method can get called frequently. I traced it all the way back to this line https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/scripts_controller.rb#L8 . So maybe something to look into?

Thanks!  I confirmed locally that we are indeed querying for all user_levels associated with a user (when the user is a student) here.  Opened https://codedotorg.atlassian.net/browse/LP-2189 to track.

**Pull request merged**